### PR TITLE
fix(transactions): Auto-scroll Activity tab when new transaction added

### PR DESCRIPTION
**Pull request by vinnyhoward**
## **Description**

When users navigate to the Activity tab after completing a send or swap transaction, the new transaction does not appear at the top of the list unless they manually scroll up. The transaction list updates with the correct data, but the view remains scrolled to its previous position, causing confusion as users don't see their latest transaction.

**Improvement/solution:**
Implemented automatic scroll-to-top functionality when new transactions are added to the Activity tab. Created a reusable `useTransactionAutoScroll` custom hook that:

- Detects when new transactions are added to the list (either list growth or first item change)
- Automatically scrolls to the top to show the latest transaction
- Prevents disruptive scrolling when users are manually browsing the list
- Works across all supported networks (EVM, Solana, Bitcoin, Tron)
- Includes robust error handling and memory leak prevention

## **Changelog**

CHANGELOG entry: Fixed transaction list not automatically scrolling to show latest transactions after send/swap operations

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/TMCU-273

## **Manual testing steps**

```gherkin
Feature: Transaction list auto-scroll on new transaction

  Scenario: user completes a transaction and views Activity tab
    Given user is on the wallet home screen
    And user has existing transactions in their Activity list
    
    When user sends or swaps tokens
    And the transaction is submitted
    And user navigates to the Activity tab
    
    Then the Activity list should automatically scroll to the top
    And the new transaction should be visible at the top of the list
    
  Scenario: user is manually scrolling when new transaction arrives
    Given user is on the Activity tab
    And user is actively scrolling through their transaction history
    
    When a new transaction is detected
    
    Then the list should NOT auto-scroll
    And user should maintain their current scroll position
    
  Scenario: user resumes browsing after manual scroll
    Given user was manually scrolling
    And user has stopped scrolling for 1 second
    
    When a new transaction is added
    
    Then the list should auto-scroll to show the new transaction
```

## **Screenshots/Recordings**

Solana

https://github.com/user-attachments/assets/eff75cc9-0202-49bd-bf7b-bf54935b96ac

EVM

https://github.com/user-attachments/assets/6c38ca39-1e59-414b-adf2-22119b9336fd

TRX

https://github.com/user-attachments/assets/4cb4045b-ba71-44e9-8b58-6dd70585bb99

### **Before**

https://github.com/user-attachments/assets/8c3a08d8-bf37-4ac0-a391-1924ab308bff

### **After**

https://github.com/user-attachments/assets/c276f244-e7b3-417b-aaa0-d2b68816e479

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Touches core transaction list scrolling behavior; while guarded against user interaction and errors, it could still introduce unexpected scroll jumps or timing issues on some devices.
> 
> **Overview**
> **Unified Activity list now auto-scrolls to the top when new transactions are added.** `UnifiedTransactionsView` wires a new `useTransactionAutoScroll` hook into the `FlashList` `onScroll` handler and uses a chain-agnostic `keyExtractor` (EVM `tx.id`, non-EVM `id` with `chain-timestamp` fallback) to detect when the first item changes or the list grows.
> 
> The new hook debounces auto-scroll with a small delay, avoids scrolling while the user is actively scrolling, and includes timeout cleanup plus error handling around ID extraction and `scrollToOffset`. A comprehensive test suite was added to cover multi-chain behavior, user-scroll suppression, configuration options, and edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e80ae844cc78463c94044a4a7cef042618281930. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    userScrollTimeoutRef.current = setTimeout(() => {
+      isUserScrollingRef.current = false;
+    }, 1000);
+  }, []);
+
+  /**
+   * Scroll to top when a new item is added
+   */
+  useEffect(() => {
+    if (!enabled) return;
```
[the remark this answers is not in the document]
Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts]: Fixed [a25cbe4](https://github.com/MetaMask/metamask-mobile/pull/25467/commits/a25cbe4416d6401e78e2c0b0fe0ca57c0674ea8f)

File: app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+   */
+  const handleScroll = useCallback(() => {
+    isUserScrollingRef.current = true;
+    if (userScrollTimeoutRef.current) {
+      clearTimeout(userScrollTimeoutRef.current);
+    }
+    userScrollTimeoutRef.current = setTimeout(() => {
+      isUserScrollingRef.current = false;
+    }, 1000);
+  }, []);
```
[the remark this answers is not in the document]
Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts:55]: Fixed [da20820](https://github.com/MetaMask/metamask-mobile/pull/25467/commits/da20820aeb4e36416f96b913f209df47fa063740)

File: app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.test.ts [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+      act(() => {
+        jest.advanceTimersByTime(150);
+      });
+
+      expect(listRef.current?.scrollToOffset).toHaveBeenCalledWith({
+        offset: 0,
+        animated: true,
+      });
+    });
+  });
```
[the remark this answers is not in the document]
Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.test.ts]: fixed [da20820](https://github.com/MetaMask/metamask-mobile/pull/25467/commits/da20820aeb4e36416f96b913f209df47fa063740)

File: app/components/Views/UnifiedTransactionsView/UnifiedTransactionsView.tsx [hunk 4]
```diff
@@ -541,6 +542,17 @@ const UnifiedTransactionsView = ({
...
   const listRef = useRef<FlashListRef<UnifiedItem>>(null);
 
+  // Auto-scroll to top when new transactions are added
+  const { handleScroll } = useTransactionAutoScroll(data, listRef, {
+    getItemId: (item: UnifiedItem) => {
+      if (item.kind === TransactionKind.Evm) {
+        return getTransactionId(item.tx) ?? null;
+      }
+      // For non-EVM (Solana, Bitcoin, Tron, etc.)
+      return item.tx?.id ? String(item.tx.id) : null;
```
[the remark this answers is not in the document]
Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/UnifiedTransactionsView.tsx]: fixed [da20820](https://github.com/MetaMask/metamask-mobile/pull/25467/commits/da20820aeb4e36416f96b913f209df47fa063740)

File: app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+   */
+  enabled?: boolean;
+  /**
+   * Delay in milliseconds before scrolling (to ensure list has updated)
+   */
+  delay?: number;
+  /**
+   * Function to extract a unique ID from an item
+   */
+  getItemId: (item: T) => string | null;
```
Comment (wachunei) [inline app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts]: Nit here: I would rename this option to be more explicit about what purpose it serves (extract the id from the generic type of `data`). Following React patterns like [`keyExtractor`](https://reactnative.dev/docs/flatlist#keyextractor) it could be `idExtractor` (actually the default [`keyExtractor`](https://github.com/facebook/react-native/blob/1c8f297d550bfa097790fa1ba44cb3925861e06e/packages/virtualized-lists/Lists/VirtualizeUtils.js#L248) should work)

Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts]: Got it, good call out. Will fix here in a bit

Comment (vinnyhoward) [reply app/components/Views/UnifiedTransactionsView/useTransactionAutoScroll.ts]: updated [3c63194](https://github.com/MetaMask/metamask-mobile/pull/25467/commits/3c63194439e590b744e5eedaea87a2109f048bcc)

**Review by wachunei — APPROVED** [review]
